### PR TITLE
fix(federation): Prevent @tag and @inaccessible usages on fields marked @external

### PR DESCRIPTION
**Pull request by trevor-scheer**
This PR introduces new composition errors to prevent the usage of `@tag` and `@inaccessible` on a field that is marked `@external`. Since `@tag` is part of Studio's API, we're going to block this usage for now to prevent confusion. Similarly, `@inaccessible` usage on `@external` doesn't really mean anything, so allowing it will probably lead to confusion as users may not know what to expect.

This may be relaxed in the future in order to accommodate an actual semantic usage of the combination or possibly just a warning (warnings don't exist yet, but this could be a good candidate for one if we choose to go down that path).

This PR also handles the rename of `appliedDirectives` recommended in feedback on #859.

Fixes: #867

### CODE REVIEW

File: docs/source/errors.md [hunk 1]
```diff
@@ -53,6 +53,12 @@ If Apollo Gateway encounters an error, composition fails. This document lists co
 | `REQUIRES_FIELDS_MISSING_ON_BASE` | The `fields` argument of an entity field's `@requires` directive includes a field that is not defined in the entity's originating subgraph.`|
 | `REQUIRES_USED_ON_BASE` | An entity field is marked with `@requires` in the entity's originating subgraph, which is invalid. |
 
+## `@tag`
+
+| Code | Description |
+|---|---|
+| `TAG_USED_WITH_EXTERNAL` | Fields marked as `@external` cannot also have `@tag` usages.  |
```
Comment (martijnwalraven) [inline docs/source/errors.md:60]: This seems fine for now, but I wonder if we should generalize this to other directives.

File: federation-js/src/composition/types.ts [hunk 2]
```diff
@@ -39,7 +39,7 @@ export interface FederationField {
   requires?: ReadonlyArray<SelectionNode>;
   provides?: ReadonlyArray<SelectionNode>;
   belongsToValueType?: boolean;
-  appliedDirectives?: DirectiveNode[]
+  otherKnownDirectives?: DirectiveNode[]
```
Comment (martijnwalraven) [inline federation-js/src/composition/types.ts:42]: Note that these are directive usages / applications, not directive definitions. (The naming in `graphql-js` is confusing, but `GraphQLDirective` is a directive definition, while `DirectiveNode` is a usage.)

**Review by martijnwalraven — COMMENTED** [review]
The logic looks good to me, just a small comment about naming and the difference between directive definitions and usages.

File: docs/source/errors.md (hunk 1 above)
Comment (prasek) [reply docs/source/errors.md:60]: @martijnwalraven @trevor-scheer should we add the same handling for`@inaccessible` here as well -- even if we don't generalize just yet?

Comment (prasek) [reply docs/source/errors.md:60]: also how will we handle `@tag` and `@inaccessible` on value types referenced only from `@external` fields, like https://github.com/apollographql/supergraph-demo/pull/165/files#diff-4a1edface097b589a87522e4fc919429284c343628e3bca83170a6db5e1bc678R9?

Comment (trevor-scheer) [reply docs/source/errors.md:60]: @prasek Included inaccessible in this PR, see commit 3d51e58

Going to handle value type concerns in a separate PR for the sake of timing.

File: federation-js/src/composition/types.ts (hunk 2 above)
Comment (trevor-scheer) [reply federation-js/src/composition/types.ts]: Got it! Went through and clarified the naming a bit further - everything is either named usage or definition accordingly. A bit verbose but helpful I think.

**Review by martijnwalraven — APPROVED** [review]
Assuming we'll figure out handling of value types in a follow up PR, this looks good to me!

**Comment by trevor-scheer** [comment]
Overridden by #880

**Pull request closed without merging**